### PR TITLE
Feature: Add campaign pages

### PR DESCRIPTION
**Pull request by kjohnson**
Resolves [GIVE-1150]

## Description

This PR introduces the Campaign Page, which can be customized via the Block Editor.

From the campaign list table (coming soon), the campaign page can be accessed via `/wp-admin/?action=edit_campaign_page&campaign_id={id}`.

The Campaign Page model is created "on demand" when attempting to customize the page content (see the testing instructions). Ideally, this provides us flexibility to update the "default" state of the page when not customized (as blocks can be complicated to update if created ahead of time).

Note: This does not include front-end routing of the Campaign for publicly rendering the page content.

## Affects

The Campaign/Page relationship has been moved to the CampaignPage class, which now stores the related `campaignId` as a property (as opposed to the Campaign having a `pageId` property). This prevents broken references to a CampaignPage when the underlying post is deleted or trashed.

## Visuals

image

## Testing Instructions

With a Campaign created (manually or via the database), go to `/wp-admin/?action=edit_campaign_page&campaign_id={id}` to be redirected to the related campaign page in the Block Editor.

## Pre-review Checklist

-   [x] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@unreleased` tags included in DocBlocks
-   [x] Includes unit tests
-   [ ] Reviewed by the designer (if follows a design)
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

[GIVE-1150]: https://stellarwp.atlassian.net/browse/GIVE-1150?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/Campaigns/ServiceProvider.php [hunk 1]
```diff
@@ -34,6 +34,9 @@ public function boot(): void
         $this->registerMenus();
         $this->registerActions();
         $this->registerMigrations();
+
```
Comment (alaca) [inline src/Campaigns/ServiceProvider.php:37]: Let's put these in the `registerActions` method - since it already exists.

**Review by alaca — APPROVED** [review]
Looks good to me 👍

File: src/Campaigns/Actions/RegisterCampaignPagePostType.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ */
+class RegisterCampaignPagePostType
+{
+    /**
+     * @unreleased
+     */
+    public function __invoke()
+    {
+        register_post_type( 'give_campaign_page', [
+            'label' => __('Campaign Page', 'give-peer-to-peer'),
```
Comment (glaubersilva) [inline src/Campaigns/Actions/RegisterCampaignPagePostType.php:16]: Let's update this textdomain

File: src/Campaigns/Repositories/CampaignPageRepository.php [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+                    'meta_key' => 'campaignId',
+                    'meta_value' => $campaignPage->campaignId,
+                ]);
+
+        } catch (Exception $exception) {
+            DB::query('ROLLBACK');
+
+            Log::error('Failed creating a campaign page', [$campaignPage]);
+
+            throw new $exception('Failed creating a donation');
```
Comment (glaubersilva) [inline src/Campaigns/Repositories/CampaignPageRepository.php:89]: Let's replace "donation" with "campaign page".

**Review by glaubersilva — COMMENTED** [review]
@kjohnson Great work! Just left a couple of comments related to small string tweaks.

File: src/Campaigns/ServiceProvider.php (hunk 1 above)
Comment (kjohnson) [reply src/Campaigns/ServiceProvider.php]: I extracted a new `setupCampaignPages()` method, see [a0abd60](https://github.com/impress-org/givewp/pull/7510/commits/a0abd6061d5c29950a9bf0f95fb46f3be0407f21).

File: src/Campaigns/Actions/RegisterCampaignPagePostType.php (hunk 2 above)
Comment (kjohnson) [reply src/Campaigns/Actions/RegisterCampaignPagePostType.php]: Resolved, see [466f978](https://github.com/impress-org/givewp/pull/7510/commits/466f978ca0313b47cd25a11d4c387527f8320cd9).

File: src/Campaigns/Repositories/CampaignPageRepository.php (hunk 3 above)
Comment (kjohnson) [reply src/Campaigns/Repositories/CampaignPageRepository.php]: Resolved, see [466f978](https://github.com/impress-org/givewp/pull/7510/commits/466f978ca0313b47cd25a11d4c387527f8320cd9).

**Pull request merged**